File: backend/src/test/java/cz/forgottenempire/servermanager/steamauth/SteamAuthControllerUnitTest.java
package cz.forgottenempire.servermanager.steamauth;

import cz.forgottenempire.servermanager.api.model.AuthStatus;
import cz.forgottenempire.servermanager.api.model.AuthType;
import cz.forgottenempire.servermanager.api.model.AuthVerificationResultDto;
import cz.forgottenempire.servermanager.api.model.SteamAuthDto;
import cz.forgottenempire.servermanager.api.model.SteamAuthStatusDto;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class SteamAuthControllerUnitTest {

    @Mock
    private SteamAuthService steamAuthService;
    @Mock(stubOnly = true)
    private SteamAuthVerifier steamAuthVerifier;

    private SteamAuthController controller;

    @BeforeEach
    void setUp() {
        controller = new SteamAuthController(steamAuthService, steamAuthVerifier);
    }

    @Test
    void whenGetAuthAccount_thenRespondWithAuthAccountWithoutPassword() {
        SteamAuth steamAuth = new SteamAuth(1L, "username", "password", "DEFGH");
        when(steamAuthService.getAuthAccount()).thenReturn(steamAuth);

        ResponseEntity<SteamAuthDto> response = controller.getSteamAuth();

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotNull();
        assertThat(response.getBody().getUsername()).isEqualTo("username");
        assertThat(response.getBody().getSteamGuardToken()).isEqualTo("DEFGH");
        assertThat(response.getBody().getPassword()).isNull();
    }

    @Test
    void whenSetAuthAccount_thenSteamAuthServiceCalled() {
        SteamAuthDto dto = new SteamAuthDto();
        dto.setUsername("username");
        dto.setPassword("password");
        dto.setSteamGuardToken("DEFGH");

        ResponseEntity<?> response = controller.setSteamAuth(dto);

        verify(steamAuthService).setAuthAccount(dto);
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNull();
    }

    @Test
    void whenClearAuthAccount_thenAuthServiceClearMethodCalledAndNoContentResponseEntityReturned() {
        ResponseEntity<?> response = controller.clearSteamAuth();

        verify(steamAuthService).clearAuthAccount();
        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);
        assertThat(response.getBody()).isNull();
    }

    @Test
    void whenGetAuthStatus_andAuthIsConfigured_thenReturnTrueStatus() {
        when(steamAuthService.isAuthConfigured()).thenReturn(true);

        ResponseEntity<SteamAuthStatusDto> response = controller.getSteamAuthStatus();

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotNull();
        assertThat(response.getBody().getIsConfigured()).isTrue();
    }

    @Test
    void whenGetAuthStatus_andAuthIsNotConfigured_thenReturnFalseStatus() {
        when(steamAuthService.isAuthConfigured()).thenReturn(false);

        ResponseEntity<SteamAuthStatusDto> response = controller.getSteamAuthStatus();

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotNull();
        assertThat(response.getBody().getIsConfigured()).isFalse();
    }

    @Test
    void whenVerifyCredentials_andVerificationSucceeds_thenReturnVerificationResult() {
        SteamAuthDto authDto = new SteamAuthDto();
        authDto.setUsername("username");
        authDto.setPassword("password");

        AuthVerificationResult verificationResult = AuthVerificationResult.builder()
                .status(AuthVerificationResult.AuthStatus.SUCCESS)
                .authType(AuthVerificationResult.AuthType.NONE)
                .message("Authentication successful")
                .build();

        when(steamAuthVerifier.verifyCredentials(authDto)).thenReturn(verificationResult);

        ResponseEntity<AuthVerificationResultDto> response = controller.verifySteamAuth(authDto);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotNull();
        assertThat(response.getBody().getStatus()).isEqualTo(AuthStatus.SUCCESS);
        assertThat(response.getBody().getAuthType()).isEqualTo(AuthType.NONE);
        assertThat(response.getBody().getMessage()).isEqualTo("Authentication successful");
    }

    @Test
    void whenVerifyCredentials_andVerificationThrowsException_thenReturnErrorResult() {
        SteamAuthDto authDto = new SteamAuthDto();
        authDto.setUsername("username");
        authDto.setPassword("password");

        when(steamAuthVerifier.verifyCredentials(authDto)).thenThrow(new RuntimeException("Test exception"));

        ResponseEntity<AuthVerificationResultDto> response = controller.verifySteamAuth(authDto);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR);
        assertThat(response.getBody()).isNotNull();
        assertThat(response.getBody().getStatus()).isEqualTo(AuthStatus.ERROR);
        assertThat(response.getBody().getMessage()).contains("Test exception");
        assertThat(response.getBody().getAuthType()).isEqualTo(AuthType.UNKNOWN);
    }
}

File: backend/src/test/kotlin/cz/forgottenempire/servermanager/steamauth/SteamAuthControllerUnitTest.kt
package cz.forgottenempire.servermanager.steamauth

import cz.forgottenempire.servermanager.api.model.AuthType
import cz.forgottenempire.servermanager.api.model.SessionStatus
import cz.forgottenempire.servermanager.api.model.SteamAuthDto
import cz.forgottenempire.servermanager.api.model.SteamAuthStatusDto
import cz.forgottenempire.servermanager.api.model.SteamLoginRequestDto
import cz.forgottenempire.servermanager.api.model.SteamLoginResult
import cz.forgottenempire.servermanager.api.model.SteamLoginResultDto
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity

@ExtendWith(MockitoExtension::class)
class SteamAuthControllerUnitTest {

    @Mock
    private lateinit var steamAuthService: SteamAuthService

    @Mock
    private lateinit var loginService: SteamLoginService

    @Mock
    private lateinit var sessionStatusHolder: SteamSessionStatusHolder

    private lateinit var controller: SteamAuthController

    @BeforeEach
    fun setUp() {
        controller = SteamAuthController(steamAuthService, loginService, sessionStatusHolder)
    }

    @Test
    fun `when get auth account, then respond with username without password`() {
        val steamAuth = SteamAuth(1L, "username", "password")
        `when`(steamAuthService.getAuthAccount()).thenReturn(steamAuth)

        val response: ResponseEntity<SteamAuthDto> = controller.getSteamAuth()

        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body).isNotNull
        assertThat(response.body!!.username).isEqualTo("username")
        assertThat(response.body!!.password).isNull()
    }

    @Test
    fun `when clear auth account, then auth service clear method called and no content response returned`() {
        val response = controller.clearSteamAuth()

        verify(steamAuthService).clearAuthAccount()
        assertThat(response.statusCode).isEqualTo(HttpStatus.NO_CONTENT)
        assertThat(response.body).isNull()
    }

    @Test
    fun `when get auth status and auth is configured, then return configured with session status`() {
        `when`(steamAuthService.isAuthConfigured()).thenReturn(true)
        `when`(sessionStatusHolder.status).thenReturn(SessionStatus.ACTIVE)

        val response: ResponseEntity<SteamAuthStatusDto> = controller.getSteamAuthStatus()

        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body!!.isConfigured).isTrue()
        assertThat(response.body!!.sessionStatus).isEqualTo(SessionStatus.ACTIVE)
    }

    @Test
    fun `when get auth status and auth is not configured, then return NOT_CONFIGURED session status`() {
        `when`(steamAuthService.isAuthConfigured()).thenReturn(false)

        val response: ResponseEntity<SteamAuthStatusDto> = controller.getSteamAuthStatus()

        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body!!.isConfigured).isFalse()
        assertThat(response.body!!.sessionStatus).isEqualTo(SessionStatus.NOT_CONFIGURED)
    }

    @Test
    fun `when steam login succeeds, then return SUCCESS result`() {
        val request = SteamLoginRequestDto().apply { username = "username"; password = "password" }
        `when`(loginService.login("username", "password", null))
            .thenReturn(SteamLoginServiceResult(SteamLoginResult.SUCCESS, AuthType.NONE, "Login successful."))

        val response: ResponseEntity<SteamLoginResultDto> = controller.steamLogin(request)

        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body!!.result).isEqualTo(SteamLoginResult.SUCCESS)
    }

    @Test
    fun `when steam login needs email code, then return CODE_REQUIRED with EMAIL auth type`() {
        val request = SteamLoginRequestDto().apply { username = "username"; password = "password" }
        `when`(loginService.login("username", "password", null))
            .thenReturn(SteamLoginServiceResult(SteamLoginResult.CODE_REQUIRED, AuthType.EMAIL, "Check your email."))

        val response: ResponseEntity<SteamLoginResultDto> = controller.steamLogin(request)

        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body!!.result).isEqualTo(SteamLoginResult.CODE_REQUIRED)
        assertThat(response.body!!.authType).isEqualTo(AuthType.EMAIL)
    }

    @Test
    fun `when steam login service throws exception, then return ERROR result`() {
        val request = SteamLoginRequestDto().apply { username = "username"; password = "password" }
        `when`(loginService.login("username", "password", null)).thenThrow(RuntimeException("Test exception"))

        val response: ResponseEntity<SteamLoginResultDto> = controller.steamLogin(request)

        assertThat(response.statusCode).isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR)
        assertThat(response.body!!.result).isEqualTo(SteamLoginResult.ERROR)
        assertThat(response.body!!.message).contains("Test exception")
    }
}
